rclpy/rclpy/executors.py

```diff
@@ -732,18 +733,22 @@ class MultiThreadedExecutor(Executor):
     Runs callbacks in a pool of threads.
 
     :param num_threads: number of worker threads in the pool. If ``None``, the number of threads
-        will use :func:`multiprocessing.cpu_count`. If that's not implemented the number of threads
-        defaults to 1.
+        will use :func:`multiprocessing.cpu_count` (minimum of 2).
+        If that's not implemented, the number of threads defaults to 2.
     :param context: The context associated with the executor.
     """
 
     def __init__(self, num_threads: int = None, *, context: Context = None) -> None:
         super().__init__(context=context)
         if num_threads is None:
             try:
-                num_threads = multiprocessing.cpu_count()
+                num_threads = max(multiprocessing.cpu_count(), 2)
             except NotImplementedError:
-                num_threads = 1
+                num_threads = 2
```

Comment:
Looking at this closer, I think we should actually refactor this a bit.

First, on Linux it is more appropriate to use `len(os.get_schedaffinity(0))`, which will tell you the number of cores this process has access to, not just how many are in the machine (see https://docs.python.org/3/library/os.html#os.sched_getaffinity and https://docs.python.org/3/library/multiprocessing.html#multiprocessing.cpu_count).  That said, `os.get_schedaffinity` doesn't work on either macOS or Windows, so we need another solution there.

But that solution doesn't need to import the multiprocessing library.  In particular, we can use `os.cpu_count()` (which is what `multiprocessing.cpu_count()` calls anyway.  The only difference between the two is that os.cpu_count() returns None if it can't work, while `multiprocessing.cpu_count()` throws an exception.

So with all of that said, and the change you want to make, I think we should make this:

```
if num_threads is None:
     # On Linux, os.sched_getaffinity() returns the number of cores this process has access to.
     # On all other platforms, os.sched_getaffinity() doesn't exist so we use the number of cores.
     if hasattr(os, 'sched_getaffinity'):
        num_threads = len(os.sched_getaffinity(0))
    else:
        num_threads = os.cpu_count()

    # The calls above may still return None if they aren't supported; in that case, just use a hard-coded 2.
    if num_threads is None:
        num_threads = 2
```

(and with that change, we can then remove the `import multiprocessing` at the top of the file)

Reply:
Thanks for pointing that out, sounds reasonable to me.

If the CPU affinity is set, it should be honored here as well.
I am not sure `multiprocessing.cpu_count()` or `os.cpu_count()` collects offline (fallback) cpu counts, but using `len(os.sched_getaffinity(0))` at 1st place does make sense to me.

I will fix it according, i will borrow some of your code! thanks 👍

Reply:
i think follow-up required for `rclcpp` as well.

Reply:
and another question is should we honor the affinity set by user? it the user set the affinity to only one CPU, we should honor that setting, so minimum can be 1.

Reply:
> I am not sure `multiprocessing.cpu_count()` or `os.cpu_count()` collects offline (fallback) cpu counts, but using `len(os.sched_getaffinity(0))` at 1st place does make sense to me.

As far as I understand, `cpu_count()` ends up calling [sysconf(_SC_NPROCESSORS_ONLN)](https://github.com/python/cpython/blob/87b5fd9d9de11c68b9940d3a1dba902138129109/Modules/posixmodule.c#L13382), so it should only collect online CPUs.

> i think follow-up required for `rclcpp` as well.

That's true, it would be better to make this change for `rclcpp` as well.  It's going to be more work there, as we are going to have to invoke platform-specific logic to get it.  Probably we should implement a function down in `rcutils` to get this information, and then use that in rclcpp.

> and another question is should we honor the affinity set by user? it the user set the affinity to only one CPU, we should honor that setting, so minimum can be 1.

Yes, the new code should do that, at least on Linux.  That is, `os.sched_getaffinity` will return 1 if the user set the affinity to only one CPU, and then they will get the warning that this isn't a good idea.  But it should still run.  On Windows and macOS, I have no idea how to do this.